setup_hooks.sh

```diff
@@ -0,0 +1,36 @@
+#!/bin/sh
+rustup component add rustfmt-preview
+
+rustfmt_path=`find $HOME/.rustup/toolchains/ -wholename '*bin/rustfmt*'`
+echo "#!/bin/sh
+declare -a rust_files=()
+declare -a cpp_files=()
+
+files=\$(git diff-index --name-only HEAD)
+echo 'Formatting source files'
+for file in \$files; do
+    if [ ! -f \"\${file}\" ]; then
+        continue
+    fi
+    if [[ \"\${file}\" == *.rs ]]; then
+        rust_files+=(\"\${file}\")
+    fi
+    if [[ \"\${file}\" =~ (\.h|\.cpp|\.cc) ]]; then
+        cpp_files+=(\"\${file}\")
+    fi
+done
+echo \${rust_files[@]} | xargs --no-run-if-empty $rustfmt_path &
+echo \${cpp_files[@]} | xargs --no-run-if-empty clang-format -i &
```

Comment:
`--no-run-if-empty` is only available for GNU toolchain, that is, not available on Mac OS X....

Reply:
Do you know any alternative? I'd really live to run these tasks in parallel.

Reply:
This is verbose, but helpful for dealing with OS specific issues...:

```
if [ -n "$files" ]; then
    echo $files | xargs &
fi
```

Reply:
Yea, but how could it be piped in `&` parallel execution then?

Reply:
if [ -n ${rust_files[@]} ]; then
     echo ${rust_files[@]} | xargs $rustfmt_path &
fi
if [ -n ${rust_files[@]} ]; then
     echo ${cpp_files[@]} | xargs clang-format -i &
fi

Reply:
Hmm, i'm not bash pro. Does it work? This syntax seems to be broken to me.

Reply:
Ah, I see. `&` is not a conditional operator, it just starts a new job. Ok

Reply:
Just updated.

Explanation: if the array has zero element, this line returns exit code 0; otherwise, because `||` is or, we run the second half, it is the same as what you have, except -P 4, meaning, at most four parallelizations.

I suggest removing `&` and `wait` to just make the format process in the foreground.

Reply:
Edit again: reading `xargs` manual, it seems that `-P 4` does not seem necessary.

Reply:
See edit. `xargs` is not needed anymore